x/tally/keeper/gas_meter_test.go

```diff
@@ -0,0 +1,214 @@
+
+		reveals := map[string]types.RevealBody{
+			"a": {ExitCode: 0, Reveal: `{"result": {"text": "A"}}`, GasUsed: g0, ProxyPubKeys: proxyPubKeys},
+			"b": {ExitCode: 0, Reveal: `{"result": {"text": "A"}}`, GasUsed: g1, ProxyPubKeys: proxyPubKeys},
+			"c": {ExitCode: 0, Reveal: `{"result": {"text": "A"}}`, GasUsed: g2, ProxyPubKeys: proxyPubKeys},
+			"d": {ExitCode: 0, Reveal: `{"result": {"text": "A"}}`, GasUsed: g3, ProxyPubKeys: proxyPubKeys},
+			"e": {ExitCode: 0, Reveal: `{"result": {"text": "A"}}`, GasUsed: g4, ProxyPubKeys: proxyPubKeys},
+			"f": {ExitCode: 0, Reveal: `{"result": {"text": "A"}}`, GasUsed: g5, ProxyPubKeys: proxyPubKeys},
+			"g": {ExitCode: 0, Reveal: `{"result": {"text": "A"}}`, GasUsed: g6, ProxyPubKeys: proxyPubKeys},
+			"h": {ExitCode: 0, Reveal: `{"result": {"text": "A"}}`, GasUsed: g7, ProxyPubKeys: proxyPubKeys},
+			"i": {ExitCode: 0, Reveal: `{"result": {"text": "A"}}`, GasUsed: g8, ProxyPubKeys: proxyPubKeys},
+			"j": {ExitCode: 0, Reveal: `{"result": {"text": "A"}}`, GasUsed: g9, ProxyPubKeys: proxyPubKeys},
+		}
+
+		expExecGasUsed, expTotalExecGasUsed := expectedExecutorGas(reveals, totalProxyGasUsed, execGasLimit, uint64(len(reveals)))
+
+		gasMeter := types.NewGasMeter(tallyGasLimit, execGasLimit, types.DefaultMaxTallyGasLimit, gasPrice, types.DefaultGasCostBase)
+		_, tallyRes := fixture.tallyKeeper.FilterAndTally(
+			fixture.Context(),
+			types.Request{
+				Reveals:           reveals,
+				ReplicationFactor: uint16(len(reveals)),
+				ConsensusFilter:   base64.StdEncoding.EncodeToString(filterInput),
+				GasPrice:          gasPriceStr,
+				ExecGasLimit:      execGasLimit,
+				TallyGasLimit:     tallyGasLimit,
+				TallyProgramID:    hex.EncodeToString(tallyProgram.Hash),
+			}, types.DefaultParams(), gasMeter)
+
+		// Check executor gas used.
+		sumExec := math.NewInt(0)
+		for _, exec := range gasMeter.Executors {
+			require.Equal(t,
+				expExecGasUsed[exec.PublicKey],
+				exec.Amount.Uint64(),
+			)
+			sumExec = sumExec.Add(exec.Amount)
+		}
+		require.Equal(t, expTotalExecGasUsed, tallyRes.ExecGasUsed)
+
+		// Check proxy gas used.
+		for _, proxy := range gasMeter.Proxies {
+			require.Equal(t,
+				expProxyGasUsed[hex.EncodeToString(proxy.PublicKey)].String(),
+				proxy.Amount.String(),
+			)
+			sumExec = sumExec.Add(proxy.Amount)
+		}
+
+		sumExec = sumExec.Add(math.NewIntFromUint64(gasMeter.RemainingExecGas()))
+		require.Equal(t, sumExec.String(), strconv.FormatUint(execGasLimit, 10))
+
+		tallySum := math.NewIntFromUint64(gasMeter.TallyGasUsed())
+		tallySum = tallySum.Add(math.NewIntFromUint64(gasMeter.RemainingTallyGas()))
+		require.Equal(t, tallySum.String(), strconv.FormatUint(tallyGasLimit, 10))
+	})
+}
+
+// expectedExecutorGas returns the expected executor gas consumption and the
+// total executor gas used given the reveals.
+func expectedExecutorGas(reveals map[string]types.RevealBody, proxyGasUsed, execGasLimit, rf uint64) (map[string]uint64, uint64) {
```

Comment:
I think we touched on this briefly, but I don't see a lot of value in testing the output against basically the same implementation 😅 I think the fuzzer should mainly tell us if certain inputs cause a panic, or if the end state after a bunch of fuzzed actions is where we expect it to be.

Reply:
Yeah I agree. I’ll remove individual gas checks before merging